search.1
.Dd November 12, 2010
.Dt SEARCH 1
.Os
.Sh NAME
.Nm search
.Nd walk through a file hierarchy
.Sh SYNOPSIS
.Pp
.Nm
.Op Fl L | Fl P
.Op Fl EIsxv
.Ar path ...
.Op Fl f Ar path
.Op Fl n Ar pattern
.Op Fl r Ar regexp
.Op Fl t Ar ftype
.Op ...
.Pp
.Nm
.Op Fl L | Fl P
.Op Fl EIsxv
.Fl f Ar path
.Op path ...
.Op Fl n Ar pattern
.Op Fl r Ar regexp
.Op Fl t Ar ftype
.Op ...
.Sh DESCRIPTION
The
.Nm
will walk through a file hierachy.
.Pp
Short options:
.Bl -tag -width indent
.It Fl E
Use modern regular expressions rather than `basic' (obsolete)
regular expressions which is the default. See the man page of
.Xr re_format 7
for more information.
.It Fl I
Case insensitive search, otherwise the specified pattern is
interpreted as case sensitive.
.It Fl L
Follow symbolic links and return the information of the files
they reference. It is an error if the referenced files do not
exist.
.It Fl P
Do not follow symbolic links, but return the information the
symbolic links themselves, this is the default behaviour.
.It Fl s
Cause
.Nm
to alphabetically sort the search results. Note:
.Ql search -s
has identical output as
.Ql search | sort .
.It Fl x
Prevents
.Nm
from descending into directories that have a device
number different than that of the file from which
the descent began.
.It Fl v
Show version number
.It Fl n Ar pattern
Specify a
.Ar pattern
to match the search results. Special shell pattern matching
characters may be used as part of the pattern. See the man
page of
.Xr fnmatch 3
for more information.
.It Fl r Ar regexp
Specify a regular expression
.Ar regexp
to match the search results. Unless the option
.Ic -E
is explicitly specified, the regular expression will only be
interpreted as
.Ql basic
(obsolete) regular expressions. See the man page of
.Xr re_format 7
for more information.
.It Fl t Ar ftype
Specify a file type
.Ar ftype
to match the search results.
.Pp
.Bl -tag -width indent -compact
.It Cm b
block device
.It Cm c
character special
.It Cm d
directory
.It Cm p
fifo
.It Cm l
symbolic link
.It Cm f
regular file
.It Cm s
socket
.El
.El
.Pp
Long options:
.Bl -tag -width indent
.It Fl -delete
Delete files or directories.
.It Fl -empty
Find empty files or directories.
.It Fl -sort
Same as
.Ic -s .
.It Fl -version
Same as
.Ic -v .
.It Fl -xdev
Same as
.Ic -x .
.It Fl -gid Ar gname
Same as
.Fl -group Ar gname .
.It Fl -group Ar gname
Find files belonging to
.Ar gname .
.It Fl -uid Ar uname
Same as
.Fl -user Ar uname .
.It Fl -user Ar uname
Find files belonging to
.Ar uname .
.It Fl -nogroup
Find files that belong to an unknown group
.It Fl -nouser
Find files that belong to an unknown user
.It Fl -name Ar pattern
Same as
.Ic -n Ar pattern .
.It Fl -path Ar path
Same as
.Ic -f Ar path .
.It Fl -regex Ar regexp
Same as
.Ic -r Ar regexp .
.It Fl -type Ar ftype
Same as
.Ic -t Ar ftype .
.El
.Sh NOTES
As
.Nm
is just a programming practice work. Now it uses a doubly linked list
to sort the search results, which may consume more memory and be a bit
slower for a large file hierachy than the classic
.Ql find
found on most UNIX systems and clones.
.Sh HISTORY
A
.Nm
command was written mainly for programming practice.
.Sh AUTHOR
Denise H. G. <darcsis AT gmail DOT com>
.Sh BUGS
There must be plenty of them...


search.c
/*
 * Copyright (c) 2005-2010 Denise H. G. <darcsis@gmail.com>
 * All rights reserved
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 */

#include <getopt.h>

#include "search.h"
  
extern char *optarg;
extern int optind;

extern int init_plan(plan_t *);
extern int find_plan(int, char **, plan_t *);
extern int execute_plan(plan_t *);
extern int add_plan(plan_t *);
extern void free_plan(plist_t **);

static int opt_empty;
static int opt_delete;

static __inline void ftype_err(const char *);
static __inline void cleanup(int);

static struct option longopts[] = {
	{ "gid",     required_argument, NULL,        2  },
	{ "group",   required_argument, NULL,        3  },
	{ "path",    required_argument, NULL,       'f' },
	{ "name",    required_argument, NULL,       'n' },
	{ "regex",   required_argument, NULL,       'r' },
	{ "type",    required_argument, NULL,       't' },
	{ "user",    required_argument, NULL,        4  },
	{ "uid",     required_argument, NULL,        5  },
	{ "empty",   no_argument,       &opt_empty,  1  },
	{ "delete",  no_argument,       &opt_delete, 1  },
	{ "sort",    no_argument,       NULL,       's' },
	{ "nogroup", no_argument,       NULL,        6  },
	{ "nouser",  no_argument,       NULL,        7  },
	{ "version", no_argument,       NULL,       'v' },
	{ "xdev",    no_argument,       NULL,       'x' },
	{ NULL,      0,                 NULL,        0  }
  };

plan_t plan;

int
main(int argc, char *argv[])
{
  int ch;
  int ret;
    
  (void)setlocale(LC_CTYPE, "");
  signal(SIGINT, cleanup);

  if (init_plan(&plan) < 0) {
#ifdef _DEBUG_
	warnx("init_plan() failed!");
#endif
	cleanup(0);
	exit (1);
  }

  while ((ch = getopt_long(argc, argv, "EILPsvxf:n:r:t:", longopts, NULL)) != -1)
	switch (ch) {
	case 2:
	case 3:
	  plan.flags |= OPT_GRP;
	  bzero(plan.args->sgid, LINE_MAX);
	  strlcpy(plan.args->sgid, optarg, LINE_MAX);
	  break;
	case 4:
	case 5:
	  plan.flags |=  OPT_USR;
	  bzero(plan.args->suid, LINE_MAX);
	  strlcpy(plan.args->suid, optarg, LINE_MAX);
	  break;
	case 6:
      plan.flags |= OPT_NGRP;
      plan.flags |= OPT_IDS;
      break;
	case 7:
      plan.flags |= OPT_NUSR;
      plan.flags |= OPT_IDS;
      break;
	case 'f':
	  plan.flags |= OPT_PATH;
	  dl_append(optarg, plan.paths);
	  break;
	case 'n':
	  if (!(plan.flags & OPT_NAME)) {
		if (plan.flags & OPT_REGEX)
		  plan.flags &= ~OPT_REGEX;
		plan.flags |= OPT_NAME;
	  }
	  bzero(plan.mt->pattern, LINE_MAX);
	  strlcpy(plan.mt->pattern, optarg, LINE_MAX);
	  break;
	case 'r':
	  if (!(plan.flags & OPT_REGEX)) {
		if (plan.flags & OPT_NAME)
		  plan.flags &= ~OPT_NAME;
		plan.flags |= OPT_REGEX;
	  }
	  bzero(plan.mt->pattern, LINE_MAX);
	  strlcpy(plan.mt->pattern, optarg, LINE_MAX);
	  break;
	case 0:
	  if (opt_empty == 1)
		plan.flags |= OPT_EMPTY;
	  if (opt_delete == 1) {
		plan.flags |=  OPT_DEL;
		if ((plan.rfiles = dl_init()) == NULL ||
			(plan.rdirs = dl_init()) == NULL) {
		  warnx("error initiating lists for deletion!");
		  return (-1);
		}
	  }
	  break;
	case 's':
	  plan.flags |= OPT_SORT;
      plan.args->need_sort = 1;
	  break;
	case 'v':
	  plan.flags |= OPT_VERSION;
	  break;
	case 'x':
	  plan.flags |= OPT_XDEV;
      plan.args->need_xdev = 1;
	  break;
	case 't':
	  plan.flags |= OPT_TYPE;
	  switch (optarg[0]) {
	  case 'p':
		plan.args->type = NT_ISFIFO;
		break;
	  case 'c':
		plan.args->type = NT_ISCHR;
		break;
	  case 'd':
		plan.args->type = NT_ISDIR;
		break;
	  case 'b':
		plan.args->type = NT_ISBLK;
		break;
	  case 'l':
		plan.args->type = NT_ISLNK;
		break;
	  case 's':
		plan.args->type = NT_ISSOCK;
		break;
	  case 'f':
	  case '\0':
		plan.args->type = NT_ISREG;
		break;
	  default:
		ftype_err(optarg);
		cleanup(0);
		exit (1);
	  }
	  break;
	case 'E':
	  plan.mt->mflag |= REG_EXTENDED;
	  break;
	case 'I':
	  plan.mt->mflag |= REG_ICASE;
	  break;
	case 'L':
	  if (plan.flags & OPT_LSTAT) {
		plan.flags &= ~OPT_LSTAT;
		plan.flags |= OPT_STAT;
	  }
	  break;
	case 'P':
	  if (plan.flags & OPT_STAT) {
		plan.flags &= ~OPT_STAT;
		plan.flags |= OPT_LSTAT;
	  }
	  break;
	default:
	  plan.flags |= OPT_USAGE;
	  break;
	}

  argc -= optind;
  argv += optind;

  if (find_plan(argc, argv, &plan) < 0) {
#ifdef _DEBUG_
	warnx("find_plan() failed!");
#endif
	cleanup(0);
	exit (1);
  }
  
  if (add_plan(&plan) < 0) {
#ifdef _DEBUG_
	warnx("add_plan() failed!");
#endif
	cleanup(0);
	exit (1);
  }

  ret = execute_plan(&plan);

#ifdef _DEBUG_
  warnx("ret=%d", ret);
#endif

  cleanup(0);
  return (ret);
}

static __inline void
ftype_err(const char *s)
{
  if (s == NULL)
	return;

  warnx("--type: %s: unknown type", s);
  return;
}

static __inline void
cleanup(int sig)
{
  if (plan.plans) {
	free_plan(&(plan.plans));
	plan.plans = NULL;
  }
  
  if (plan.paths) {
	dl_free(plan.paths);
	plan.paths = NULL;
  }

  if (plan.rfiles) {
	dl_free(plan.rfiles);
	plan.rfiles = NULL;
  }

  if (plan.rdirs) {
	dl_free(plan.rdirs);
	plan.rdirs = NULL;
  }

  if (plan.mt != NULL) {
	free(plan.mt);
	plan.mt = NULL;
  }
  
  if (plan.args != NULL) {
	free(plan.args);
	plan.args = NULL;
  }

  if (plan.nstat != NULL) {
	free(plan.nstat);
	plan.nstat = NULL;
  }
  
  if (sig) {
	(void)fprintf(stderr, "\n");
	exit(0);
  }
}
